source/slang/slang-ir-autodiff-unzip.h

```diff
@@ -372,29 +372,31 @@ struct DiffUnzipPass
             }
             else
             {
-                // Ignore `out` parameters since these are removed from propagate call.
-                if (!as<IROutType>(resolvedPrimalFuncType->getParamType(ii)))
+                if (auto inOutType = as<IRInOutType>(resolvedPrimalFuncType->getParamType(ii)))
                 {
-                    if (auto inOutType = as<IRInOutType>(resolvedPrimalFuncType->getParamType(ii)))
-                    {
-                        // For 'inout' parameter we need to create a temp var to hold the value
-                        // before the primal call.
-                        //
-                        auto tempPrimalVar = primalBuilder->emitVar(as<IRPtrTypeBase>(arg->getDataType())->getValueType());
+                    // For 'inout' parameter we need to create a temp var to hold the value
+                    // before the primal call. This logic is similar to the 'inout' case for differentiable params
+                    // only we don't need to deal with pair types.
+                    //
+                    auto tempPrimalVar = primalBuilder->emitVar(as<IRPtrTypeBase>(arg->getDataType())->getValueType());
 
-                        auto storeUse = findUniqueStoredVal(cast<IRVar>(arg));
-                        auto storeInst = cast<IRStore>(storeUse->getUser());
-                        auto storedVal = storeInst->getVal();
+                    auto storeUse = findUniqueStoredVal(cast<IRVar>(arg));
+                    auto storeInst = cast<IRStore>(storeUse->getUser());
+                    auto storedVal = storeInst->getVal();
 
-                        primalBuilder->emitStore(tempPrimalVar, storedVal);
+                    primalBuilder->emitStore(tempPrimalVar, storedVal);
 
-                        diffArgs.add(tempPrimalVar);
-                    }
-                    else
-                    {
-                        // Pure 'in' type. Simply re-use the original argument inst.
-                        diffArgs.add(arg);
-                    }
+                    diffArgs.add(tempPrimalVar);
+                }
+                else
+                {
+                    // For pure 'in' type. Simply re-use the original argument inst.
```

Comment:
what if it is `out` type?

Reply:
In this case, it looks like the old out-type detection case was never hit (because it was testing the arg-type and not the param type), and instead we had logic in transposition to discard the arguments corresponding to `out` types. 

If we discard the out type args at the unzipping step, it seems to trigger some argument count equality checks. I think the right place to be doing this is at the transposition step anyway.

Reply:
Sounds good.